tests/core/test_common_utils.py

```diff
@@ -240,3 +240,59 @@ def test_populate_replay_buffer_collect_trajectories():
     )
     assert buffer.num_stored == num_trials * _MOCK_TRAJ_LEN
     assert len(buffer.trajectory_indices) == num_trials
+
+
+def test_get_basic_buffer_iterators():
+    buffer = mbrl.util.replay_buffer.ReplayBuffer(1000, (1,), (1,))
+    dummy = np.ones(1)
+    for i in range(900):
+        buffer.add(dummy, dummy, dummy, i, False)
+
+    train_iter, val_iter = mbrl.util.common.get_basic_buffer_iterators(buffer, 32, 0.1)
+    assert train_iter.num_stored == 810 and val_iter.num_stored == 90
+    all_rewards = []
+    for it in [train_iter, val_iter]:
+        for batch in it:
+            _, _, _, reward, _ = batch.astuple()
+            all_rewards.extend(reward)
+    assert sorted(all_rewards) == list(range(900))
+
+
+def test_get_sequence_buffer_iterators():
+    buffer = mbrl.util.replay_buffer.ReplayBuffer(
+        1000, (1,), (1,), max_trajectory_length=20
+    )
+    dummy = np.ones(1)
+    num_trajectories_train = 27
+    num_trajectories_val = 3
+    k = 0
+    for i in range(num_trajectories_train):
+        for j in range(20):
+            buffer.add(dummy, dummy, dummy, k, False)
+            k += 1
+        buffer.close_trajectory()
+    val_cutoff = k
+    for i in range(num_trajectories_val):
+        for j in range(20):
+            buffer.add(dummy, dummy, dummy, k, False)
+            k += 1
+        buffer.close_trajectory()
+
+    train_iter, val_iter = mbrl.util.common.get_sequence_buffer_iterator(
+        buffer, 32, 0.1, 10, 3
+    )
+    # For trajectories of length 20 and sequence length 10, there are
+    # 10 possible start states.
+    # There are 30 trajectories in total, so 10% is 3 trajectories
```

Comment:
I really like this version, much better than what I did. I am already starting to make the changes in my PR. I think there might be a logical error. If you assume a trajectory from 0-19 with length 20, then there would be 11 valid starts from 0 to 10 (inclusive) and therefore 10 + 1 valid sequences.

This test also fails if the sequences are of length one.
```python
def test_get_sequence_buffer_iterators_with_length_one():  # edge case
    buffer = mbrl.util.replay_buffer.ReplayBuffer(
        1000, (1,), (1,), max_trajectory_length=1
    )
    dummy = np.ones(1)
    num_trajectories_train = 27
    num_trajectories_val = 3
    k = 0
    for i in range(num_trajectories_train):
        buffer.add(dummy, dummy, dummy, k, False)
        buffer.close_trajectory()

    for i in range(num_trajectories_val):
        buffer.add(dummy, dummy, dummy, k, False)
        buffer.close_trajectory()

    train_iter, val_iter = mbrl.util.common.get_sequence_buffer_iterator(
        buffer, 32, 0.1, 1, 3
    )

    assert len(train_iter._valid_starts) == 27  # fails in current implementation since length is 0
    assert len(val_iter._valid_starts) == 3  # fails in current implementation
```
This is simply fixed with
```python
valid_starts.extend(list(range(start, end - sequence_length + 1)))  # line 260 in replay_buffer.py
```
I think the other tests may need to be revised

Reply:
Oh right, current version is missing that `range()`'s end index is not inclusive. The funny thing is that I wrote the other test and counted 11 possible start states (instead of 10) but when my test failed I convinced myself that 10 was correct :) Good catch!